src/python/foglamp/update_statistics_history.py

```diff
@@ -0,0 +1,139 @@
+    Returns:
+        list of distinct keys
+    """
+    key_list = []
+    stmt = sqlalchemy.select([_STATS_TABLE.c.key.distinct()]).select_from(_STATS_TABLE)
+    result = _CONN.execute(stmt)
+    result = result.fetchall()
+    for i in range(len(result)):
+        key_list.append(result[i][0].replace(" ", ""))
+
+    return key_list
+
+# List of PRIMARY KEY values in statistics
+_STATS_KEY_VALUE_LIST = __list_stats_keys()
+
+
+def __insert_into_stats_history(key='', history_ts=time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(time.time())),
+                                value=0):
+    """
+    INSERT values in statistics_history
+    Args:
+        key: corresponding stats_key_value
+        history_ts: timestamp in statistics table for corresponding key 
+        value: delta between `value` and `prev_val`
+    Returns:
+        Return the number of rows inserted. Since each process inserts only 1 row, the expected count should always 
+        be 1. 
+    """
+    stmt = _STATS_HISTORY_TABLE.insert().values(key=key, history_ts=history_ts, value=value)
+    _CONN.execute(stmt)
+
+
+def __update_previous_value(key=''):
+    """Query: 
+    UPDATE 
+        statistics 
+    SET 
+        previous_value = (SELECT value FROM statistics WHERE key='READINGS')
+    WHERE 
+        value = (SELECT value FROM statistics WHERE key='READINGS'); 
+    """
+    sub_select_stmt = sqlalchemy.select([_STATS_TABLE.c.value]).select_from(_STATS_TABLE).where(
+        _STATS_TABLE.c.key == key)
+    result = _CONN.execute(sub_select_stmt)
+    result = int(result.fetchall()[0][0])
+    update_stmt = _STATS_TABLE.update().values(previous_value=result).where(key == key)
+    _CONN.execute(update_stmt)
+
+
+def __select_from_statistics(key='') -> dict:
+    """
+    SELECT data from statistics for the statistics_history table
+    Args:
+        key: The row name update is executed against (WHERE condition)
+
+    Returns:
+
+    """
+    stmt = sqlalchemy.select([_STATS_TABLE.c.value, _STATS_TABLE.c.previous_value,
+                              _STATS_TABLE.c.ts]).where(_STATS_TABLE.c.key == key)
+
+    result = _CONN.execute(stmt)
+    return result.fetchall()
+
+
+def stats_history_main():
+    """
+    1. SELECT against the  stats table,
+    2. INSERT into history table, setting value to be the delta between stats.value and stats.previous_value 
+    3. UPDATE the stats table with new values
+    Returns:
+
+    """
+    for key in _STATS_KEY_VALUE_LIST:
+        stats_select_result = __select_from_statistics(key=key)
+        value = int(stats_select_result[0][0])
+        previous_value = int(stats_select_result[0][1])
+        history_ts = time.mktime(stats_select_result[0][2].timetuple())
+        history_ts = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(history_ts))
+
+        __insert_into_stats_history(key=key, history_ts=history_ts, value=value-previous_value)
+        __update_previous_value(key=key)
```

Comment:
update_previous_value should have the value passed into it as an argument
__update_previous_value(key=key, value=value)